nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java

```diff
@@ -2617,4 +2617,248 @@ public void orElseLibraryModelSupport() {
+            "  public void test1() {",
+            "    b = new Item();",
+            "  }",
+            "  @EnsuresNonNull({\"b\", \"a\"})",
+            "  public void test2() {",
+            "    super.test2();",
+            "    b = new Item();",
+            "  }",
+            "  // BUG: Diagnostic contains: postcondition inheritance is violated, this method must guarantee that all fields written in overridden method @EnsuresNonNull annotation are @NonNull at exit point as well. Fields [a] must explicitly appear as parameters at this method @EnsuresNonNull annotation",
+            "  public void noAnnotation() {",
+            "  }",
+            "}")
+        .addSourceLines(
+            "Item.java", "package com.uber;", "class Item {", "  public void call() { }", "}")
+        .doTest();
+  }
+
+  @Test
+  public void supportEnsuresAndRequiresNonNullContract() {
+    compilationHelper
+        .addSourceLines(
+            "Content.java",
+            "package com.uber;",
+            "import javax.annotation.Nullable;",
+            "import com.uber.nullaway.qual.RequiresNonNull;",
+            "import com.uber.nullaway.qual.EnsuresNonNull;",
+            "class Content {",
+            "  @Nullable Item nullItem;",
+            "  @RequiresNonNull(\"nullItem\")",
+            "  public void run() {",
+            "    nullItem.call();",
+            "  }",
+            "  @EnsuresNonNull(\"nullItem\")",
+            "  public void init() {",
+            "    nullItem = new Item();",
+            "  }",
+            "  public void test1() {",
+            "    init();",
+            "    run();",
+            "  }",
+            "  public void test2() {",
+            "    Content content = new Content();",
+            "    content.init();",
+            "    content.run();",
+            "  }",
+            "  public void test3() {",
+            "    Content content = new Content();",
+            "    init();",
+            "    Content other = new Content();",
+            "    other.init();",
+            "    // BUG: Diagnostic contains: expected field [nullItem] is not non-null at call site",
+            "    content.run();",
+            "  }",
+            "}")
+        .addSourceLines(
+            "Item.java", "package com.uber;", "class Item {", "  public void call() { }", "}")
+        .addSourceLines(
+            "Box.java",
+            "package com.uber;",
+            "class Box {",
+            "  Content content = new Content();",
+            "}")
+        .addSourceLines(
+            "Office.java",
+            "package com.uber;",
+            "class Office {",
+            "  Box box = new Box();",
+            //            "  public void test() {",
```

Comment:
Why is this commented out?

Reply:
I am so sorry, that was a mistake, It's uncommented now.

Reply:
Thank you for great comments, I think this is ready for another review.